[14.0][MIG] account_fiscal_year_closing

SimoRubi opened the pull request:
Migration from https://github.com/OCA/account-closing/tree/60de43819512b8b2df30dff2fb5cc4bda0b11fce

antespi committed:
[ADD] account_fiscalyear_closing: Generic engine for closing operations

pedrobaeza committed:
[FIX+IMP] account_fiscal_year_closing

* Included template mechanism
* Remove unused files
* Fix behaviors
* Some refactoring
* Complete features

JordiBForgeFlow committed:
migrate account_fiscal_year_closing to 10.0

oca-transbot committed:
OCA Transbot updated translations from Transifex

sergiocorato committed:
[12.0][PORT] account fiscal year closing

oca-travis committed:
[UPD] Update account_fiscal_year_closing.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-closing-12.0/account-closing-12.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-12-0/account-closing-12-0-account_fiscal_year_closing/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-closing-12.0/account-closing-12.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-12-0/account-closing-12-0-account_fiscal_year_closing/

alexey-pelykh committed:
[FIX] account_fiscal_year_closing: tests should not use fixed dates

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
account_fiscal_year_closing 12.0.1.0.1

eLBati committed:
Translated using Weblate (Italian)

Currently translated at 66.6% (76 of 114 strings)

Translation: account-closing-12.0/account-closing-12.0-account_fiscal_year_closing
Translate-URL: https://translation.odoo-community.org/projects/account-closing-12-0/account-closing-12-0-account_fiscal_year_closing/it/

eLBati committed:
IMP account_fiscal_year_closing adding multi company rule for account.fiscalyear.closing.template

OCA-git-bot committed:
account_fiscal_year_closing 12.0.1.1.0

marcelsavegnago committed:
Added translation using Weblate (Portuguese (Brazil))

SimoRubi committed:
[MIG] account_fiscal_year_closing: Migration to 14.0
Fixed warnings:
Field account.fiscalyear.closing.closing_template_id: unknown parameter 'oldname', if this is an actual parameter you may want to override the method _valid_field_parameter on the relevant model in order to allow it
Field account.move.fyc_id: unknown parameter 'delete', if this is an actual parameter you may want to override the method _valid_field_parameter on the relevant model in order to allow it
The model account.fiscalyear.closing.config has no _description
The model account.fiscalyear.closing.mapping has no _description
The model account.fiscalyear.closing.type has no _description
The model account.fiscalyear.closing.template has no _description
The model account.fiscalyear.closing.config.template has no _description
The model account.fiscalyear.closing.mapping.template has no _description
The model account.fiscalyear.closing.type.template has no _description
The model account.fiscalyear.closing.unbalanced.move has no access rules
The model account.fiscalyear.closing.unbalanced.move.line has no access rules
odoo.addons.base.models.ir_ui_view: Search tag requires at least one field element
The method '_company_default_get' on res.company is deprecated and shouldn't be used anymore
DeprecationWarning: RedirectWarning method 'post()' is a deprecated alias to 'action_post()' or _post()

SimoRubi committed:
[IMP] account_fiscal_year_closing: black, isort, prettier

sergiocorato commented on account_fiscal_year_closing/__manifest__.py:
```diff
@@ -0,0 +1,26 @@
+    "version": "14.0.1.0.0",
+    "category": "Accounting & Finance",
+    "website": "https://github.com/OCA/account-closing",
+    "author": "Tecnativa, " "Odoo Community Association (OCA)",
```
Is this `" "` admitted?

SimoRubi replied:
Thanks for having a look!
Yes it's admitted apparently (there are no complaints by linters) but it's ugly, just removed it

sergiocorato commented on account_fiscal_year_closing/models/account_fiscalyear_closing.py:
```diff
@@ -0,0 +1,701 @@
+        }
+
+    # @api.onchange('closing_template_id')
+    # Not working due to https://github.com/odoo/odoo/issues/20163
```
Is this still valid?

SimoRubi replied:
Apparently not: https://github.com/odoo/odoo/issues/20163#issuecomment-863133625.
I enabled the onchange and removed both the method `action_load_template` and `is_new_template` that existed only to perform the onchange's job.

Tried the loading on onchange in my local environment and seems to be working fine

sergiocorato commented on account_fiscal_year_closing/models/account_move.py:
```diff
@@ -0,0 +1,33 @@
+    closing_type = fields.Selection(
+        selection=_selection_closing_type,
+        default="none",
+        states={"posted": [("readonly", True)]},
```
A value here should always exist, see #165

SimoRubi replied:
I'd rather wait for https://github.com/OCA/account-closing/pull/165 to be merged and then eventually re-base to have the merged changes

petrus-v commented:
I'm not sure if it's related to my data or something else I'm getting the following traceback while clicking on calculate button:

**As far I understand given a quick analysis** I'm not sure if we should add the date on the model `account.fiscalyear.closing.unbalanced.move.line` or remove it in the [move_prepare](https://github.com/OCA/account-closing/blob/d530593d7f70e0f2c2fd81cca8048b1e19d3a5b9/account_fiscal_year_closing/models/account_fiscalyear_closing.py#L454) ?

```
Erreur:
Odoo Server Error

Traceback (most recent call last):
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 359, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 347, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 912, in __call__
    return self.method(*args, **kw)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 531, in response_wrap
    response = f(*args, **kw)
  File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/web/controllers/main.py", line 1393, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/web/controllers/main.py", line 1381, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 396, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 383, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 328, in button_calculate
    res = self.calculate()
  File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 315, in calculate
    return self._show_unbalanced_move_wizard(data)
  File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 295, in _show_unbalanced_move_wizard
    wizard = self.env["account.fiscalyear.closing.unbalanced.move"].create(data)
  File "<decorator-gen-65>", line 2, in create
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 344, in _model_create_multi
    return create(self, [arg])
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/addons/base/models/ir_fields.py", line 533, in create
    recs = super().create(vals_list)
  File "<decorator-gen-13>", line 2, in create
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 345, in _model_create_multi
    return create(self, arg)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/models.py", line 3868, in create
    records = self._create(data_list)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/models.py", line 4028, in _create
    for other, data in zip(others, data_list)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/fields.py", line 3038, in create
    self.write_batch(record_values, True)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/fields.py", line 3064, in write_batch
    return self.write_real(records_commands_list, create)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/fields.py", line 3236, in write_real
    flush()
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/fields.py", line 3200, in flush
    comodel.create(to_create)
  File "<decorator-gen-65>", line 2, in create
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 345, in _model_create_multi
    return create(self, arg)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/addons/base/models/ir_fields.py", line 533, in create
    recs = super().create(vals_list)
  File "<decorator-gen-13>", line 2, in create
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 345, in _model_create_multi
    return create(self, arg)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/models.py", line 3825, in create
    raise ValueError("Invalid field %r on model %r" % (key, self._name))
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 639, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 315, in _handle_exception
    raise exception.with_traceback(None) from new_cause
ValueError: Invalid field 'date' on model 'account.fiscalyear.closing.unbalanced.move.line'
```

SimoRubi committed:
[IMP] account_fiscal_year_closing: Restore onchange on closing's template

SimoRubi committed:
[IMP] account_fiscal_year_closing: Add date field to unbalanced move line.
Otherwise the following might happen:
  File "/path/to/odoo/addons/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 328, in button_calculate
    res = self.calculate()
  File "/path/to/odoo/addons/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 315, in calculate
    return self._show_unbalanced_move_wizard(data)
  File "/path/to/odoo/addons/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 295, in _show_unbalanced_move_wizard
    wizard = self.env["account.fiscalyear.closing.unbalanced.move"].create(data)
[...]
ValueError: Invalid field 'date' on model 'account.fiscalyear.closing.unbalanced.move.line'

SimoRubi commented:
Hi, thanks for having a look!

> I'm not sure if it's related to my data or something else I'm getting the following traceback while clicking on calculate button:
> 
> **As far I understand given a quick analysis** I'm not sure if we should add the date on the model `account.fiscalyear.closing.unbalanced.move.line` or remove it in the [move_prepare](https://github.com/OCA/account-closing/blob/d530593d7f70e0f2c2fd81cca8048b1e19d3a5b9/account_fiscal_year_closing/models/account_fiscalyear_closing.py#L454) ?
> 
> ```
> Erreur:
> Odoo Server Error
> 
> Traceback (most recent call last):
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
>     result = request.dispatch()
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 683, in dispatch
>     result = self._call_function(**self.params)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 359, in _call_function
>     return checked_call(self.db, *args, **kwargs)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/service/model.py", line 94, in wrapper
>     return f(dbname, *args, **kwargs)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 347, in checked_call
>     result = self.endpoint(*a, **kw)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 912, in __call__
>     return self.method(*args, **kw)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 531, in response_wrap
>     response = f(*args, **kw)
>   File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/web/controllers/main.py", line 1393, in call_button
>     action = self._call_kw(model, method, args, kwargs)
>   File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/web/controllers/main.py", line 1381, in _call_kw
>     return call_kw(request.env[model], method, args, kwargs)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 396, in call_kw
>     result = _call_kw_multi(method, model, args, kwargs)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 383, in _call_kw_multi
>     result = method(recs, *args, **kwargs)
>   File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 328, in button_calculate
>     res = self.calculate()
>   File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 315, in calculate
>     return self._show_unbalanced_move_wizard(data)
>   File "/home/pverkest/clients/cgi/cgi-erp/addons-contrib/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 295, in _show_unbalanced_move_wizard
>     wizard = self.env["account.fiscalyear.closing.unbalanced.move"].create(data)
>   File "<decorator-gen-65>", line 2, in create
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 344, in _model_create_multi
>     return create(self, [arg])
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/addons/base/models/ir_fields.py", line 533, in create
>     recs = super().create(vals_list)
>   File "<decorator-gen-13>", line 2, in create
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 345, in _model_create_multi
>     return create(self, arg)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/models.py", line 3868, in create
>     records = self._create(data_list)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/models.py", line 4028, in _create
>     for other, data in zip(others, data_list)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/fields.py", line 3038, in create
>     self.write_batch(record_values, True)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/fields.py", line 3064, in write_batch
>     return self.write_real(records_commands_list, create)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/fields.py", line 3236, in write_real
>     flush()
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/fields.py", line 3200, in flush
>     comodel.create(to_create)
>   File "<decorator-gen-65>", line 2, in create
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 345, in _model_create_multi
>     return create(self, arg)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/addons/base/models/ir_fields.py", line 533, in create
>     recs = super().create(vals_list)
>   File "<decorator-gen-13>", line 2, in create
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/api.py", line 345, in _model_create_multi
>     return create(self, arg)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/models.py", line 3825, in create
>     raise ValueError("Invalid field %r on model %r" % (key, self._name))
> Exception
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 639, in _handle_exception
>     return super(JsonRequest, self)._handle_exception(exception)
>   File "/home/pverkest/clients/cgi/cgi-erp/ocb/odoo/http.py", line 315, in _handle_exception
>     raise exception.with_traceback(None) from new_cause
> ValueError: Invalid field 'date' on model 'account.fiscalyear.closing.unbalanced.move.line'
> ```

The field `date` the error is mentioning does not look like it is assigned in [move_prepare](https://github.com/OCA/account-closing/blob/d530593d7f70e0f2c2fd81cca8048b1e19d3a5b9/account_fiscal_year_closing/models/account_fiscalyear_closing.py#L454) because it should be in the lines, and there are several places where this might happen like `move_line_prepare`, or `dest_move_line_prepare`.

I haven't managed to reproduce your error but seems reasonable to add the `date` field, done in last commit.
Can you check?

SimoRubi commented:
@sergiocorato can you update your review? Unfortunately #165 hasn't been merged yet so I think there is nothing to change here for https://github.com/OCA/account-closing/pull/169#discussion_r631642120.

@petrus-v any news about https://github.com/OCA/account-closing/pull/169#issuecomment-863217963? The issue you found should be now fixed.

petrus-v commented:
@SimoRubi 

I had a quick test regarding this module and sorry for this poor report without the way to reproduce that I can't even remember too...

I can see your last commit should avoid this error it could be nice to add a test case to cover it but without the way to reproduce it's a bit annoying.

To be honest I'll be pretty busy in next few weeks and not that much interest here to give a try right now :-/

mide-dynapps commented:
Hello, 
I found a little bug on "Calculate" process:
```
  File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 293, in button_calculate
    res = self.calculate()
  File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 277, in calculate
    move, data = config.moves_create()
  File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 514, in moves_create
    move_ids = self.inverse_move_prepare()
  File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 503, in inverse_move_prepare
    return move_ids.ids
Exception
 
The above exception was the direct cause of the following exception:
 
Traceback (most recent call last):
  File "/home/odoo/src/odoo/odoo/http.py", line 640, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/odoo/src/odoo/odoo/http.py", line 316, in _handle_exception
    raise exception.with_traceback(None) from new_cause
AttributeError: 'bool' object has no attribute 'ids'
```

In method [inverse_move_prepare ](https://github.com/SimoRubi/account-closing/blob/77ac991fd35889ed17443cf8d9e6bd4a7c5f49f1/account_fiscal_year_closing/models/account_fiscalyear_closing.py#L487) of AccountFiscalyearClosingConfig, we have the case of move_ids stay False and can't return ids.

Is it possible to have a look ? returning move_ids (w/o .ids)

Thanks for your works.

SimoRubi commented:
> Hello, I found a little bug on "Calculate" process:
> 
> ```
>   File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 293, in button_calculate
>     res = self.calculate()
>   File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 277, in calculate
>     move, data = config.moves_create()
>   File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 514, in moves_create
>     move_ids = self.inverse_move_prepare()
>   File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 503, in inverse_move_prepare
>     return move_ids.ids
> Exception
>  
> The above exception was the direct cause of the following exception:
>  
> Traceback (most recent call last):
>   File "/home/odoo/src/odoo/odoo/http.py", line 640, in _handle_exception
>     return super(JsonRequest, self)._handle_exception(exception)
>   File "/home/odoo/src/odoo/odoo/http.py", line 316, in _handle_exception
>     raise exception.with_traceback(None) from new_cause
> AttributeError: 'bool' object has no attribute 'ids'
> ```
> 
> In method [inverse_move_prepare ](https://github.com/SimoRubi/account-closing/blob/77ac991fd35889ed17443cf8d9e6bd4a7c5f49f1/account_fiscal_year_closing/models/account_fiscalyear_closing.py#L487) of AccountFiscalyearClosingConfig, we have the case of move_ids stay False and can't return ids.
> 
> Is it possible to have a look ? returning move_ids (w/o .ids)
> 
> Thanks for your works.

Hi, thanks for testing this.
Could you please provide the steps you performed in order to have this error?

mide-dynapps commented:
> > Hello, I found a little bug on "Calculate" process:
> > ```
> >   File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 293, in button_calculate
> >     res = self.calculate()
> >   File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 277, in calculate
> >     move, data = config.moves_create()
> >   File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 514, in moves_create
> >     move_ids = self.inverse_move_prepare()
> >   File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 503, in inverse_move_prepare
> >     return move_ids.ids
> > Exception
> >  
> > The above exception was the direct cause of the following exception:
> >  
> > Traceback (most recent call last):
> >   File "/home/odoo/src/odoo/odoo/http.py", line 640, in _handle_exception
> >     return super(JsonRequest, self)._handle_exception(exception)
> >   File "/home/odoo/src/odoo/odoo/http.py", line 316, in _handle_exception
> >     raise exception.with_traceback(None) from new_cause
> > AttributeError: 'bool' object has no attribute 'ids'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In method [inverse_move_prepare ](https://github.com/SimoRubi/account-closing/blob/77ac991fd35889ed17443cf8d9e6bd4a7c5f49f1/account_fiscal_year_closing/models/account_fiscalyear_closing.py#L487) of AccountFiscalyearClosingConfig, we have the case of move_ids stay False and can't return ids.
> > Is it possible to have a look ? returning move_ids (w/o .ids)
> > Thanks for your works.
> 
> Hi, thanks for testing this. Could you please provide the steps you performed in order to have this error?

I'm not an accounting expert, we are testing a v14 coming from a v12.
I have theses 4 configs in the system:
![image](https://user-images.githubusercontent.com/79934758/137146210-c7b29880-dbda-4dbb-bb5b-f73fde865a7c.png)

And during the 'Calculate' action, the error is raised.

mide-dynapps commented:
Hello, I come back after deep investigation.
In fact, my previous issue was linked to the "Moves configuration", inside, no "Accounts mapping" was correctly created.
I decided to delete and recreate entirely the fiscal year closing from scratch.

### 1. Issue on save
My templates are filled correctly (each config have the "accounts mapping", except the latest):
![image](https://user-images.githubusercontent.com/79934758/137805416-234332e8-46dc-4419-80a6-ac0c9a70f799.png)

_How to reproduce:_
Create new "Fiscal year closing" and select the template.
All move configurations are loaded in screen.
Save
```
odoo.sql_db: bad query: INSERT INTO "account_fiscalyear_closing_mapping" ("id", "create_uid", "create_date", "write_uid", "write_date", "dest_account_id", "fyc_config_id", "name", "src_accounts") VALUES (nextval('account_fiscalyear_closing_mapping_id_seq'), 2, (now() at time zone 'UTC'), 2, (now() at time zone 'UTC'), ARRAY[4530, '0499998 Risultato dell''esercizio'], 41, 'Chiusura Conto Economico Costi', '03%') RETURNING id
ERROR: invalid input syntax for integer: "0499998 Risultato dell'esercizio"
LINE 1: ...UTC'), 2, (now() at time zone 'UTC'), ARRAY[4530, '0499998 R...
                                                             ^
...

Traceback (most recent call last):
  File "/home/odoo/src/odoo/odoo/http.py", line 640, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/odoo/src/odoo/odoo/http.py", line 316, in _handle_exception
    raise exception.with_traceback(None) from new_cause
psycopg2.DataError: invalid input syntax for integer: "0499998 Risultato dell'esercizio"
LINE 1: ...UTC'), 2, (now() at time zone 'UTC'), ARRAY[4530, '0499998 R...
```
![image](https://user-images.githubusercontent.com/79934758/137806686-60c9a35f-bd17-4414-88e0-885950dc44c1.png)
The onchange result returns a list for **dest_account_id** and the save is broken.

I found this issue which talks about this kind of issue: https://github.com/odoo/odoo/issues/11236#issuecomment-274690847
_(the many2one inside a one2many fails)_
I applied theses changes in the code of class **AccountFiscalyearClosingMapping** by surcharging create/write methods
```
    @api.model
    def create(self, vals):
        if 'dest_account_id' in vals:
            vals['dest_account_id'] = vals['dest_account_id'][0]
        res = super(AccountFiscalyearClosingMapping, self).create(vals)
        return res

    def write(self, vals):
        if 'dest_account_id' in vals:
            vals['dest_account_id'] = vals['dest_account_id'][0]
        res = super(AccountFiscalyearClosingMapping, self).write(vals)
        return res
```
And the save works again !

### 2. Warning on deprecated **force_company**:
just change by: tmpl = self.closing_template_id.with_company(self.company_id.id)

```
 odoo.models: Context key 'force_company' is no longer supported. Use with_company(company) instead. 
 ....
 File "/home/odoo/src/user/account_fiscal_year_closing/models/account_fiscalyear_closing.py", line 206, in onchange_template_id
     tmpl = self.closing_template_id.with_context(force_company=self.company_id.id)
   File "/home/odoo/src/odoo/odoo/models.py", line 5110, in with_context
     stack_info=True,
```


I don't know if you have theses issues.
With theses fixes, all seems work fine.

Regards.

francesco-ooops commented:
@SimoRubi hi, is there a plan/roadmap for completing porting of this module?
Thanks!

SimoRubi commented:
> @SimoRubi hi, is there a plan/roadmap for completing porting of this module? Thanks!

Hi, this is in my TODO list but I don't know when I'll have time to spend for this.

If you want to work on this you are welcome to supersede just leave a note here and create a new PR starting from this branch.

elvise commented:
hi @sergiocorato any chance for your review ?

francesco-ooops commented:
@OCA/accounting-maintainers any chance to review this porting?

matteoopenf commented:
@OCA/accounting-maintainers  review?

elvise commented:
> @OCA/accounting-maintainers review?

@matteoopenf there is a pending review for @sergiocorato

matteoopenf commented:
> @api.model
>     def create(self, vals):
>         if 'dest_account_id' in vals:
>             vals['dest_account_id'] = vals['dest_account_id'][0]
>         res = super(AccountFiscalyearClosingMapping, self).create(vals)
>         return res
> 
>     def write(self, vals):
>         if 'dest_account_id' in vals:
>             vals['dest_account_id'] = vals['dest_account_id'][0]
>         res = super(AccountFiscalyearClosingMapping, self).write(vals)
>         return res

https://github.com/OCA/account-closing/pull/199 Superseed @SimoRubi  @[francesco-ooops](https://github.com/francesco-ooops)  @[mde-spring](https://github.com/mde-spring) @sergiocorato

francesco-ooops commented:
@SirTakobi

matteoopenf commented:
@SirTakobi la pr  ho sistemato e fixato i test di la fammi sapere
